### ISSUE TITLE
Saving content of Rich Text 

### ISSUE THREAD

**Issue by IcarusKR** [TomasMikula/RichTextFX#334]
Is there an easy way to save the formated text and reload it?

I have been checking out the example and it exactly what I need, but cant figure out how to get the formated text and save it to some file or database, and reload it again.

**Comment by afester** [comment]
You can check https://github.com/TomasMikula/RichTextFX/pull/315 where I have implemented a load/save functionality for the rich text editor example. There are still some issues which need to be solved, but it should give you an idea on how to start.

@TomasMikula From an API side, one thing I was missing is the possibility to build the model by calling methods such as `addParagraph()`, `addSegment()` and the like, on the model layer **only**, and then use the resulting `StyledDocument` and create a `StyledTextArea` with that model (or set the model on an existing `StyledTextArea`). As far as I understand, `StyledTextArea.append()` and its variants is currently the only possibility to put content into the styled text area.

In summary, it should be possible to

 * Build the model, using classes from the `model `package only. Result is a `StyledDocument`.
 * Create a `StyledTextArea` and set the previously created `StyledDocument` on it.
 * Get back the current document from the StyledTextArea and traverse its structure to create any kind of external representation.

**Comment by JordanMartinez** [comment]
@afester just to clarify, by 'model,' do you mean the `StyledDocument` (i.e. the content) and not the `StyledTextAreaModel`? 

> From an API side, one thing I was missing is the possibility to build the model by calling methods such as addParagraph(), addSegment() and the like..... In summary, it should be possible to build the model, using classes from the model package only. Result is a StyledDocument.

Were you thinking of some builder-like API for this?
````java
ReadOnlyStyledDocument.startParagraph(paragraphStyle)
    .addSegment(text, style)
    .addSegment(text, style)
    .addSegment(text, style)
    .endParagraph()
    .startParagraph(paragraphStyle)
````
or even
````java
// t(text, style) --> Tuple2<String, Style>, which is later turned into a StyledText object
ReadOnlyStyledDocument.addParagraph(paragraphStyle, t(text, style), t(text, style))
    .addParagraph(paragraphStyle, t(text, style), t(text, style))
    .addParagraph(paragraphStyle, t(text, style), t(text, style))
````

> Create a StyledTextArea and set the previously created `StyledDocument` on it.

Isn't this already possible in one of `StyledTextArea`'s constructors? By default, an `EditableStyledDocument` is created for you if you don't pass in your own. (at least for the current snapshot release)

> Get back the current document from the StyledTextArea and traverse its structure to create any kind of external representation.

What do you have in mind here?

**Comment by TomasMikula** [comment]
@IcarusKR There is no one-liner to do that. In the most general case (`StyledTextArea`), this cannot be done all automatically by `RichTextFX`, because the style objects are of unknown type (for `StyledTextArea`, they are just type parameters). It could be supported if we ask the user to provide [`Codec`s](http://fxmisc.github.io/richtext/javadoc/0.7-M1/org/fxmisc/richtext/Codec.html) to serialize the style objects and `RichTextFX` handles the rest (this is done for serializing the document to the clipboard). That, however, would create a kind of commitment for us to not change the binary format in the future&mdash;if people used it to save their documents, they might not be able to load that document with a newer version of RichTextFX, if we changed the binary format in the meantime. I think it is best for everyone if the persistent representation is under the user's control.

Here's how you can proceed:
 1. Call [`getDocument`](http://fxmisc.github.io/richtext/javadoc/0.7-M1/org/fxmisc/richtext/StyledTextArea.html#getDocument--) to obtain the content of the text area, of type `StyledDocument`, which is basically a list of `Paragraphs` (which you obtain by calling [`getParagraphs`](http://fxmisc.github.io/richtext/javadoc/0.7-M1/org/fxmisc/richtext/StyledDocument.html#getParagraphs--)). So you just need to serialize a list of paragraphs. (If you define a `Codec` for `Paragraph`, [`listCodec`](http://fxmisc.github.io/richtext/javadoc/0.7-M1/org/fxmisc/richtext/Codec.html#listCodec-org.fxmisc.richtext.Codec-) gives you a `Codec` for a list of paragraphs, but no guarantees on backwards compatibility).

        StyledDocument<PS, S> ≃ List<Paragraph<PS, S>>
 2. A `Paragraph` is basically a paragraph style description (an object of type `PS`, where `PS` is a user-defined type argument) and a list of styled text segments, so you just need to serialize those.

        Paragraph<PS, S> ≃ (PS, List<StyledText<S>>)
 3. [`StyledText`](http://fxmisc.github.io/richtext/javadoc/0.7-M1/org/fxmisc/richtext/StyledText.html) is basically just a string and a style description (an object of type `S`, which is a user-supplied type argument).

        StyledText<S> ≃ (String, S)

All in all, we have

```java
StyledDocument<PS, S> ≃ List<(PS, List<(String, S)>)>
```

You can see that it is not too hard to write your own code to serialize this.

@afester 
> * Build the model, using classes from the `model `package only. Result is a `StyledDocument`.

This is possible, although maybe we are missing some convenience methods.

> * Create a `StyledTextArea` and set the previously created `StyledDocument` on it.

See @JordanMartinez's comment above. I don't see much benefit in being able to change the underlying document of the area after construction, if that's what you mean. You can just create a new area.

> * Get back the current document from the StyledTextArea and traverse its structure to create any kind of external representation.

[StyledTextArea.getDocument()](http://fxmisc.github.io/richtext/javadoc/0.7-M1/org/fxmisc/richtext/StyledTextArea.html#getDocument--)

**Comment by JordanMartinez** [comment]
Due to lack of activity on this issue, I'm closing it. @IcarusKR If you still need help, reopen it and ask your questions.
